java-dynamic-sqs-listener-core/src/main/java/com/jashmore/sqs/broker/concurrent/ConcurrentMessageBroker.java

```diff
@@ -0,0 +1,194 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+
+/**
+ * Handles the processing of messages across a number of threads that can dynamically change during processing.
+ */
+@Slf4j
+public class ConcurrentMessageBroker extends AbstractMessageBroker {
+    private final MessageRetriever messageRetriever;
+    private final MessageProcessor messageProcessor;
+    private final ConcurrentMessageBrokerProperties properties;
+
+    public ConcurrentMessageBroker(final MessageRetriever messageRetriever,
+                                   final MessageProcessor messageProcessor,
+                                   final ExecutorService executorService,
+                                   final ConcurrentMessageBrokerProperties properties) {
+        super(executorService);
+
+        this.messageRetriever = messageRetriever;
+        this.messageProcessor = messageProcessor;
+        this.properties = properties;
+    }
+
+    @Override
+    protected Controller createController(final CompletableFuture<Object> controllerFinishedFuture) {
+        return new ConcurrentThreadController(messageRetriever, messageProcessor, properties, controllerFinishedFuture);
+    }
+
+    /**
+     * Runnable that will be executed that handles the starting and stopping of threads that process messages.
+     */
+    @VisibleForTesting
+    @RequiredArgsConstructor
+    static class ConcurrentThreadController implements Controller {
+        private final MessageRetriever messageRetriever;
+        private final MessageProcessor messageProcessor;
+        private final ConcurrentMessageBrokerProperties properties;
+
+        /**
+         * Executor that is used to start threads from the thread pool.
+         *
+         * <p>This is a specific {@link ThreadPoolExecutor} because when the controller is to be stopped we should wait for all of the currently running
+         * threads to be stopped.
+         */
+        private final ThreadPoolExecutor executor = new ThreadPoolExecutor(0, Integer.MAX_VALUE, 60L, SECONDS, new SynchronousQueue<>());
+        /**
+         * Semaphore used to control the number of threads that are available to be run.
+         *
+         * <p>This is set to zero but will be replaced by the value from the {@link #properties}.
+         */
+        private final ResizableSemaphore resizableSemaphore = new ResizableSemaphore(0);
+        /**
+         * Completable Future that should be resolved when this thread has fully finished processing.
+         */
+        private final CompletableFuture<Object> futureToResolve;
+        /**
+         * When shutting down this will determine whether any messages currently being processed should be interrupted.
+         */
+        private boolean interruptThreads;
+
+        @Override
+        @SuppressFBWarnings({"RV_RETURN_VALUE_IGNORED_BAD_PRACTICE"})
+        public void run() {
```

Comment:
Worth noting here in comment that this is the main control thread and at most one thread will be executing through here at anytime - otherwise there could be state issues.

Reply:
Yeah I need to document more the Thread confinement details, etc!